ls1intum/Artemis
# `Programming exercises`: Re-enable Haskell exercises on Jenkins setups

### Checklist
#### General


~- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.~
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
During the transition to Jenkins pipeline jobs Haskell exercises got disabled for Jenkins setups. This PR makes the feature available again.
Closes #4964 

### Description
- A new `Jenkinsfile` that defines the execution steps in Jenkins has been added.
- Haskell exercises require that the solution repository is checked out as well. The server code has been adapted to allow for the necessary replacements in the `Jenkinsfile`.
- The Haskell test runner creates slightly different test results. The Jenkins plugin has been adapted to parse this format: https://github.com/ls1intum/jenkins-server-notification-plugin/pull/19
- Jenkins does not run the Docker container as root. The existing container has been adapted accordingly (for now placed in https://github.com/b-fein/artemis-haskell).

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student

**To test it on TS2, the updated Jenkins plugin (https://github.com/ls1intum/jenkins-server-notification-plugin/pull/19) has to be deployed there first. For local setups, you can download the plugin manually from the [artefacts section of the CI job](https://github.com/ls1intum/jenkins-server-notification-plugin/actions/runs/2580997361) and install it in `<YOUR_JENKINS_URL>/pluginManager/advanced` → ‘Deploy Plugin’.**

1. Create a new Haskell programming exercise.
2. The solution should pass all tests, the template none.
3. Participate as student in the exercise. The test results should show up as usual.

### Review Progress





#### Code Review
#### Manual Tests Bamboo
#### Manual Tests Jenkins

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|

unchanged
